### DISCUSSION TITLE
Navigation Stack issue

### DISCUSSION THREAD

**Discussion by easiwriter** [pointfreeco/swift-composable-architecture#3407] [Q&A] [answered]
I've been trying to implement Stack Navigation in my AppFeature using your Stacks video and have hit a scoping issue in the AppView.

Here's the code with the error shown:

<img width="1209" alt="Screenshot 2024-09-24 at 13 58 25" src="https://github.com/user-attachments/assets/07187131-1967-417e-9742-ae110f4f7c49">

And here is the path reducer:

<img width="832" alt="Screenshot 2024-09-24 at 13 59 44" src="https://github.com/user-attachments/assets/c60a5972-e6f0-44f2-bd88-aa6498110f5b">

There is something wrong with the action parameter, but I'm blowed if I can work out what.

**Comment by mbrandonw** [comment]
Hi @easiwriter, you will need to share more code, such as what the `AppFeature` state and actions look like. You can also refer to the docs for more information on using navigation stacks.

**Reply by easiwriter** [reply]
Here is some more code. I have looked through the docs (that I can find). I'm still getting used to the fact that the library is changing and that some things get superceded.
[Archive.zip](https://github.com/user-attachments/files/17118351/Archive.zip)

**Reply by mbrandonw** [reply]
> I'm still getting used to the fact that the library is changing and that some things get superceded.

Nothing about these APIs has changed since the day they were first introduced. I'm not sure what you are referring to here.

The following code compiles fine, showing that the problem has nothing to do with the `$store.scope`:

<details>
<summary>Expand</summary>

```swift
import SwiftUI
import ComposableArchitecture

struct AppView_: View {
  @Bindable var store: StoreOf<AppFeature_>

  var body: some View {
    NavigationStack(path: $store.scope(state: \.path, action: \.path)) {
      EmptyView()
//      ProjectsListView(store: Store(initialState: ProjectsList.State()) {
//        ProjectsList()
//      })
    } destination: { store in
      switch store.state {
      case .folders:
        EmptyView()
//        if let store = store.scope(state: \.folders, actions: \.folders) {
//          FoldersListView(store: store)
//        }
      }
    }
  }
}

import ComposableArchitecture
import SwiftUI

@Reducer
struct AppFeature_: Reducer {
  @ObservableState
  struct State {
//    @MainActor
//    let db = Database(modelContext: WriteApp.sharedModelContainer.mainContext)
    var path = StackState<Path.State>()
    var projectsList = ProjectsList.State()
  }

  enum Action {
    case path(StackAction<Path.State, Path.Action>)
    case projectsList(ProjectsList.Action)
  }

  @Reducer
  struct Path {
    @ObservableState

    enum State {
      case folders(FoldersList.State)
    }

    enum Action {
      case folders(FoldersList.Action)
    }

    var body: some ReducerOf<Self> {
      Scope(state: \.folders, action: \.folders) {
        FoldersList()
      }
    }
  }

  var body: some ReducerOf<Self> {
    Scope(state: \.projectsList, action: \.projectsList) {
      ProjectsList()
    }
    Reduce { state, action in
      switch action {
      case .projectsList:
        return .none
      case .path:
        return .none
      }
    }
    .forEach(\.path, action: \.path) {
      Path()
    }
  }
}

@Reducer struct FoldersList {}
@Reducer struct ProjectsList {}
```
</details>

I commented out code until it compiled, so your problem must lie elsewhere, and you need to do the work to figure out the problem. Typically you can comment out things until it compiles and then work your way backwards.

I will say that the project you have shared has many potential problems in it. You are creating database connections in a feature's `State.init`, creating stores in views, and holding onto "view models" in state. All of that is going to lead to code with strange behavior and difficult to understand. I highly recommend reading the docs to understand more how the library works, or possibly trying to build your features with vanilla SwiftUI until you are more comfortable with how SwiftUI works.

**Reply by easiwriter** [reply]
The API I refer to is NavigationViewStore that changed to NavigationView. This is documented in the migration notes for 1.7.  I have been relying on the videos you have produced,especially Stack Navigation #246. I have also studied the example code for SyncUps. 

I have struggled to find a reference source for the APIs that one normally finds in help documents.  I am aware of the various articles but these are not always up to date as far as I can tell.

The potential problems you mention are not problems I don't think. The database connection you mention is not actually a database connection but a poorly named static reference to a set of ViewModels accessed using a container managed by Swinject. There is no View code involved and all the ViewModels are structs. I have moved the Database reference out of the Reducer. I don't know what you mean by 'creating stores in views'?

**Reply by easiwriter** [reply]
I have spent more hours trying to get NavigationStack to compile, in particular:

    NavigationStack(path: $store.scope(state: \.path,
                                       action: \.path))

I get the following errors, yet the same code in SyncUps compiles. This makes no sense.

<img width="674" alt="Screenshot 2024-09-26 at 09 59 32" src="https://github.com/user-attachments/assets/397426a6-9c12-4059-9877-5da21b80e1f1">

<img width="711" alt="Screenshot 2024-09-26 at 10 00 00" src="https://github.com/user-attachments/assets/ba2f6dad-7697-4a67-be0c-dc8bc82d151e">

I'm sure this is one of those situations where the solution is obvious, so please help.

**Reply by easiwriter** [reply]
I've just got the view to compile by removing some stuff as shown below:

<img width="1038" alt="Screenshot 2024-09-26 at 17 13 40" src="https://github.com/user-attachments/assets/1c706c60-1580-438a-b341-3c7d5ec119bb">

I don't know why but will experiment. I'm away for a week from tomorrow so will attack it again when I get back. Meanwhile any words of advice would be most welcome.

**Reply by kevinanderson7** [reply]
Hey @easiwriter, I'm not able to download your project to take a look at the moment, but based on the code you have posted above, it looks like you're scoping on the folderList when really all you should need to do is grab the associated value from the folderList case of the switch and pass it to the FolderListView directly. That store from the associated value is already scoped. Here is the code from SyncUps for reference:
```
} destination: { store in
      switch store.case {
      case let .detail(store):
        SyncUpDetailView(store: store)
      case let .meeting(meeting, syncUp):
        MeetingView(meeting: meeting, syncUp: syncUp)
      case let .record(store):
        RecordMeetingView(store: store)
      }
```

There may be other errors as well though. I'd need to see the code for RootFeature to have a better idea though.

**Reply by easiwriter** [reply, marked as answer]
Thanks Kevin. I've just resolved the problem. It was due to faulty nesting of braces and a missing case let. Stupid really, but that's the way it goes.
